crates/symbolicator/src/services/cacher.rs

```diff
@@ -308,26 +308,39 @@ impl<T: CacheItemRequest> Cacher<T> {
         let dup_file = temp_file.reopen()?;
         let mut temp_fd = tokio::fs::File::from_std(dup_file);
 
-        // TODO: consider cache expiry!
         let shared_cache_hit = self
             .shared_cache_service
             .fetch(&shared_cache_key, &mut temp_fd)
             .await;
 
-        let status = if !shared_cache_hit {
-            let status = request.compute(temp_file.path()).await?;
-            status.write(&mut temp_fd).await?;
-            Some(status)
-        } else {
-            None
+        let status = match shared_cache_hit {
+            true => {
+                // Waste an mmap call on a cold path, oh well.
+                let bv = ByteView::map_file_ref(temp_file.as_file())?;
+                request
+                    .should_load(&bv)
+                    .then(|| CacheStatus::from_content(&bv))
+            }
+            false => None,
+        };
+        if shared_cache_hit && status.is_none() {
+            log::trace!("Discarding item from shared cache {}", key);
+            metric!(counter("shared_cache.file.discarded") += 1);
+        }
+
+        let status = match status {
+            Some(status) => status,
+            None => {
+                let status = request.compute(temp_file.path()).await?;
+                status.write(&mut temp_fd).await?;
+                status
+            }
```

Comment:
Nit: I think this is just `status.unwrap_or_else(|| …)`, but it's fine like this as well.

Reply:
you need `.await` though, which you can't write in an `.unwrap_or_else()`

Reply:
Thanks, I missed that.